### PR TITLE
fix: uv loop polling when render process reuse enabled

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/22119.
Closes https://github.com/electron/electron/issues/25314.
Closes https://github.com/electron/electron/issues/23910.
Closes https://github.com/electron/electron/issues/25405.
Closes https://github.com/electron/electron/issues/25496.

Fixes an issue where some Node.js module API calls hung in the renderer process on reload when renderer process reuse was in effect. This was happening because when render process reuse was enabled, Chromium reused the same process, but the main thread was still bound to the old iocp (on windows, and respective interface for other platforms) because libuv didn't destroy it. The queue then went on to operate on this old iocp - when we reloaded, a new iocp was created but not bound to the main thread and so we would not properly be informed of events on it. This is fixed by preparing the message loop on _every_ reload when reuse is enabled with `node_bindings_->PrepareMessageLoop()`.

cc @zcbenz @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed an issue where some Node.js module API calls hung in the renderer process after reloads when render process reuse was enabled.

### CODE REVIEW

**Comment by zcbenz** [comment]
Busy looping is not a solution for listening to events. Apart from constant high CPU usage, when consuming streams it would add the delay for each chunk, making IO very slow.

**Comment by codebytere** [comment]
@zcbenz that does make sense and i don't think this is the ideal solution either - it's not clear to me what specifically is lacking in the previous approach though. Do you have any thoughts on an alternative?

**Comment by zcbenz** [comment]
I have not looked into the issue, but it seems that the issue is caused by ` GetQueuedCompletionStatus(uv_loop_->iocp)` no longer working after reloading. So I think there are few possibilities:

1. Node has created a new uv loop or iocp after reloading while we are still using the old one. (Unlikely though since other platforms are still working.)
2. Node does some initialization work to the iocp, after reloading the iocp was initialized again and stopped working.
3. After reloading Electron does some cleanup work and breaks iocp.

**Comment by codebytere** [comment]
@zcbenz determined the true underlying cause - updated the solution & PR body accordingly

**Review by nornagon — APPROVED** [review]

**Comment by nornagon** [comment]
Can we get a test?

File: spec-main/api-browser-window-spec.ts [hunk 1]
```diff
@@ -4286,6 +4286,29 @@ describe('BrowserWindow module', () => {
...
+    it('does not cause Node.js module API hangs after reload', (done) => {
+      const w = new BrowserWindow({
+        show: false,
+        webPreferences: {
+          nodeIntegration: true
+        }
+      });
+
+      let count = 0;
+      ipcMain.on('async-node-api-done', () => {
```
Comment (zcbenz) [inline spec-main/api-browser-window-spec.ts:4299]: Should remove listener after test is over.

**Review by zcbenz — COMMENTED** [review]
Test seems to be failing on mac CI:

```
Error: done() called multiple times
    at IpcMainImpl.<anonymous> (electron/spec-main/api-browser-window-spec.ts:4301:11)
    at Object.<anonymous> (electron/js2c/browser_init.js:157:8980)
```

File: spec-main/api-browser-window-spec.ts (hunk 1 above)
Comment (codebytere) [reply spec-main/api-browser-window-spec.ts:4299]: ah whoops, you're right!

**Review by zcbenz — APPROVED** [review]

**Pull request merged**